howso/utilities/utilities.py

```diff
@@ -1212,3 +1212,183 @@ def deep_update(base, updates):
+        for the diagonal values to be filled in. If `fill_diagonal` is set to false, then this
+        parameter will be ignored.
+
+    Returns
+    -------
+    Dataframe
+        Dataframe of the result.
+    """
+    if matrix.shape[0] != matrix.shape[1]:
+        raise ValueError(
+            f"Invalid matrix shape: ({matrix.shape[0]}, {matrix.shape[1]}), "
+            "matrix must be square."
+        )
+
+    # Ensures sorting
+    matrix = matrix.sort_index(axis=0)
+    matrix = matrix.sort_index(axis=1)
+
+    if isinstance(normalize_method, str) or isinstance(normalize_method, Callable):
+        normalize_method = [normalize_method]
+
+    # Default normalization methods
+    def sum_if_not_zero(row):
+        """Standard sum."""
+        row_sum = row.sum()
+        if row_sum != 0:
+            return row / row_sum
+        else:
+            warnings.warn(
+                f"Sum of row {row.name} is 0. Division cannot be performed. "
+                "Row values will remain unnormalized."
+            )
+            return row
+
+    def abs_sum_if_not_zero(row):
+        """Sum of the absolute values."""
+        row_sum = row.abs().sum()
+        if row_sum != 0:
+            return row / row_sum
+        else:
+            warnings.warn(
+                f"Absolute sum of row {row.name} is 0. Division cannot be performed. "
+                "Row values will remain unnormalized."
+            )
+            return row
+
+    def divide_by_max_abs(row):
+        """Division by the maximum absolute value."""
+        max_abs_value = row.abs().max()
+        if max_abs_value != 0:
+            return row / max_abs_value
+        else:
+            warnings.warn(
+                f"Maximum absolute value of row {row.name} is 0. Division cannot be performed. "
+                "Row values will remain unnormalized."
+            )
+            return row
+
+    if normalize:
+        # Replace all diagonal values with Nan, then put them back after normalization.
+        if ignore_diagonals_normalize:
+            diagonal_values = np.diag(matrix).copy()
+            for i in range(len(matrix)):
+                matrix.iat[i, i] = np.nan
+        for method in normalize_method:
+            if method == "relative":
+                matrix = matrix.apply(divide_by_max_abs, axis=1)  # type: ignore
+            elif method == "sum":
+                matrix = matrix.apply(sum_if_not_zero, axis=1)  # type: ignore
+            elif method == "absolute_sum":
+                matrix = matrix.apply(abs_sum_if_not_zero, axis=1)  # type: ignore
+            elif method == "feature_count":
+                matrix = matrix.apply(lambda x: x / len(matrix.columns), axis=1)  # type: ignore
+            elif callable(method):
+                matrix = matrix.apply(method, axis=1)   # type: ignore
+            else:
+                raise ValueError(
+                    f"Invalid `normalize_method` parameter value: {normalize_method}. "
```

Comment:
This message should probably mention that it could be an iterable of these things as well.